src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -208,5 +227,130 @@ static void VerifyMemberOnlyApplyToTypesOrStrings (SymbolAnalysisContext context
+				&& context.Symbol is IMethodSymbol method) {
+				VerifyMethodDamAttributesMatch (method, overriddenMethod, context);
+			}
+		}
+
+		static void VerifyMethodDamAttributesMatch (IMethodSymbol method, IMethodSymbol overriddenMethod, SymbolAnalysisContext context)
+		{
+			// Check parameters and return types for DAM annotations
+			bool methodReturnHasDam = TryGetReturnDamAttribute (method, out var methodReturnAttribute);
+			bool overriddenReturnHasDam = TryGetReturnDamAttribute (overriddenMethod, out var overriddenReturnAttribute);
+			if ((methodReturnHasDam && overriddenReturnHasDam && !DamAttributesMatch (methodReturnAttribute!, overriddenReturnAttribute!))
+				|| (methodReturnHasDam ^ overriddenReturnHasDam)) {
+				context.ReportDiagnostic (Diagnostic.Create (
+					DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnMethodReturnValueBetweenOverrides),
+					method.Locations[0], method.GetDisplayName (), overriddenMethod.GetDisplayName ()));
+			}
+
+			for (int i = 0; i < method.Parameters.Length; i++) {
+				var methodParam = method.Parameters[i];
+				var overriddenParam = overriddenMethod.Parameters[i];
+				var methodParamHasDam = TryGetParamDamAttribute (method, i, out var methodParamAttribute);
+				var overriddenParamHasDam = TryGetParamDamAttribute (overriddenMethod, i, out var overriddenParamAttribute);
+				if ((methodParamHasDam && overriddenParamHasDam && !DamAttributesMatch (methodParamAttribute!, overriddenParamAttribute!))
+					|| (methodParamHasDam ^ overriddenParamHasDam)) {
+					context.ReportDiagnostic (Diagnostic.Create (
+						DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnMethodParameterBetweenOverrides),
+						methodParam.Locations[0],
+						methodParam.GetDisplayName (), method.GetDisplayName (), overriddenParam.GetDisplayName (), overriddenMethod.GetDisplayName ()));
+				}
+			}
+
+			for (int i = 0; i < method.TypeParameters.Length; i++) {
+				var methodTypeParam = method.TypeParameters[i];
+				var overriddenTypeParam = overriddenMethod.TypeParameters[i];
+				var methodParamHasDam = methodTypeParam.TryGetAttribute (DynamicallyAccessedMembersAttribute, out var methodTypeParamAttribute);
+				var overriddenParamHasDam = overriddenTypeParam.TryGetAttribute (DynamicallyAccessedMembersAttribute, out var overriddenTypeParamAttribute);
+				if ((methodParamHasDam && overriddenParamHasDam && !DamAttributesMatch (methodTypeParamAttribute!, overriddenTypeParamAttribute!))
+					|| (methodParamHasDam ^ overriddenParamHasDam)) {
+					context.ReportDiagnostic (Diagnostic.Create (
+						DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnGenericParameterBetweenOverrides),
+						methodTypeParam.Locations[0],
+						methodTypeParam.GetDisplayName (), method.GetDisplayName (), overriddenTypeParam.GetDisplayName (), overriddenMethod.GetDisplayName ()));
+				}
```

Comment:
Would it be possible to refactor this to share the code between the four places? The condition and warnings are nearly identical.

Reply:
I tried to, and for me the slight differences in each case are just enough so that anything I tried ended up adding complexity without reducing repetition.